File: 18fourSum/fourSum.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()        
        dictionary = {key:{*group} for key, group in itertools.groupby(sorted(filter(lambda _: operator.lt(*_), itertools.product(range(len(nums)), repeat = 2)), key = lambda _: sum(map(nums.__getitem__, _))), lambda _: sum(map(nums.__getitem__, _)))}
        result = set()
        for left in range(2, len(nums) - 1):
            for right in range(left + 1, len(nums)):
                if (complement := target - sum(map(nums.__getitem__, (left, right)))) in dictionary: result.update(tuple(map(nums.__getitem__, (*_, left, right)))  for _ in dictionary.get(complement) if _[1] < left)
        return [[*_] for _ in result]

File: 18fourSum/fourSum.java
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        Arrays.sort(nums);
        final Map<Integer, List<java.util.AbstractMap.SimpleEntry<Integer, Integer>>> dictionary = IntStream.range(0, nums.length - 1).boxed().flatMap(left -> IntStream.range(left + 1, nums.length).mapToObj(right -> new java.util.AbstractMap.SimpleEntry<>(left, right))).collect(Collectors.groupingBy($ -> nums[$.getKey()] + nums[$.getValue()]));        
        final HashSet<List<Integer>> result = new HashSet<>();
        java.util.stream.IntStream.range(2, nums.length - 1).forEach(left -> java.util.stream.IntStream.range(left + 1, nums.length).forEach(right -> {
            final int complement = target - java.util.stream.IntStream.of(left, right).map($ -> nums[$]).sum();
            if (dictionary.containsKey(complement)) dictionary.get(complement).stream().filter($ -> $.getValue() < left).forEach($_ -> result.add(java.util.stream.IntStream.of($_.getKey(), $_.getValue(), left, right).map($ -> nums[$]).boxed().collect(java.util.stream.Collectors.toList())));}));
        return new ArrayList<>(result);
    }
}
